Various small tweaks

oalders opened the pull request:
- Remove some unused variables
- Make array access less likely to be confused with array slice
- Add actual speed to > MAX_SENSIBLE_SPEED warning

oalders committed:
Make array access less likely to be confused with array slice

DDvO commented on gpxconv:
```diff
@@ -511,7 +511,6 @@ while(<>) {
         $ele_miss_length += $dis;
       }
       if (!$sec && $prev_sec && $spd ne "" && $spd > 0) {
-        $calc_sec_ = $prev_sec + $dis/$spd;  # substitute time using recorded speed and distance from previous point
```
Oh, this is actually a bug. Please keep the line included,
but replace `$calc_sec_ =` by `my $calc_sec =`.

DDvO commented on gpxconv:
```diff
@@ -603,7 +602,6 @@ while(<>) {
     $i++;
     ($prev_lat, $prev_lon, $prev_ele, $prev_tim, $prev_sec, $prev_spd) = ($lat, $lon, $ele, $tim, $sec, $spd);
     $prev_avail_ele = $ele if $ele ne "";
-    $prev_avail_tim = $tim if $tim ne "";
```
Right, the variable `$prev_avail_tim` is unused, but is still referred to in a comment,
and I need to figure out how best to handle this.
So please do not delete the line, but just comment it out.

DDvO commented on gpxconv:
```diff
@@ -679,7 +679,12 @@ for(my $i=0; $i <= $#TIM; $i++) {
-      ignoring_trkpt("speed is more than ".MAX_SENSIBLE_SPEED." km/h");
+      ignoring_trkpt(
+        sprintf(
+            'speed %.2f is more than MAX_SENSIBLE_SPEED km/h',
```
NIce. Yet please use `SPD_PRECISION` instead of `%.2f`,
replace also `MAX_SENSIBLE_SPEED` by `SPD_PRECISION`,
and re-phrase the message a little, such that the line becomes
```
'speed = '.SPD_PRECISION.' km/h is more than '.SPD_PRECISION,
```

DDvO reviewed (changes requested):
Thank you for your further contribution.
I request some improvements before I can merge these commits.

oalders committed:
Fix variable name for $calc_sec and scope to the containing block

oalders committed:
Comment out unused variable $prev_avail_tim

oalders committed:
Add actual speed to > MAX_SENSIBLE_SPEED warning

DDvO reviewed (approved):
Thanks!

oalders commented:
Thanks for the review! I've addressed your comments and force-pushed the changes so that there aren't a bunch of reverted commits in this pull request.

Just for context, I've been using this script to smooth out GPS lines for open water swims, which can be notoriously bad. It has worked quite well. I've been using `WALKING` and tweaking some of the parameters like `MAX_SENSIBLE_SPEED` in order to get changes which make sense in the context of swimming, which is going to be slower than walking.

Here's a swim from earlier this week. The lines are actually not bad after filtering because it's not a point to point swim and there was a fair amount of changing direction at the end when I was running out of sunlight and staying close to the beach.

![R3KZV37WMXEJCEMJU7DW5DCW7435XSC27DQJ75IC4TTTJRWNRZ5JHVWSAIVJB75IDCIB2MOZYLWHYA2A6UHYTB7D7SME7L2M](https://user-images.githubusercontent.com/96205/189398950-75455e96-f247-4331-9b17-8bfb543a0149.png)

This shaved a 2.0 km swim down to 1.5 km, which is much closer to what it should be.

It would be nice if the script could accept command line arguments. Right now it's a bit hard for me to make changes, send pull requests and then re-apply my changes to the constants.

Also, it might be handy to run it through `perltidy`, to handle the formatting. Not sure at which point my pull requests will become annoying, so please do let me know when I've tested your patience too much. 😄

DDvO commented:
> Thanks for the review! I've addressed your comments and force-pushed the changes so that there aren't a bunch of reverted commits in this pull request.

Yeah, good not to pollute commits with reverted ones.
Yet there is a nicer way of adapting commits before they get merged: fixup commits.
`git commit --fixup <commit-id or @~n> <file(s)>` adds a delta to the specified commit,
and when merging to the target branch,`--autosquash` can be used to combine the fixups with their respective original commit.
  
> Just for context, I've been using this script to smooth out GPS lines for open water swims, which can be notoriously bad. It has worked quite well.

Nice to learn about your use case.
And pleased to hear that the simple ad-hoc smoothing algorithm works reasonably well here.

> I've been using `WALKING` and tweaking some of the parameters like `MAX_SENSIBLE_SPEED` in order to get changes which make sense in the context of swimming, which is going to be slower than walking.

Please suggest a set of parameter values for swimming, or provide a PR for them.

> It would be nice if the script could accept command line arguments. Right now it's a bit hard for me to make changes, send pull requests and then re-apply my changes to the constants.

I've just added `-drive`, `-walk`, `-cycle`, and `-fly` options, and these can be extended easily by `-swim`.

> Also, it might be handy to run it through `perltidy`, to handle the formatting.

Good to learn about this tool - I've applied it, and then tweaked the result manually.

> Not sure at which point my pull requests will become annoying,

No worries - I'm glad to receive further improvements!

DDvO commented:
> It would be nice if the script could accept command line arguments.

I've just added a `-smooth` CLI option (and cleaned up the smoothing code).

> Right now it's a bit hard for me to make changes, send pull requests and then re-apply my changes to the constants.

What I do when I need to (re-) apply local changes is:
`git diff >changes.diff`
`patch -p1 -R changes.diff`
do other stuff
`patch -p1 changes.diff`

oalders commented:
> I've just added a -smooth CLI option (and cleaned up the smoothing code).

Thanks! I will test that out this week.

> What I do when I need to (re-) apply local changes is:
git diff >changes.diff
patch -p1 -R changes.diff
do other stuff
patch -p1 changes.diff

Right. That works too. I don't mind making patches or stashing or using fixup commits, but if the things are configurable via command line switches or a config file, that's just so much easier. 😄

oalders commented:
> Please suggest a set of parameter values for swimming, or provide a PR for them.

I will do that this week.

> I've just added -drive, -walk, -cycle, and -fly options, and these can be extended easily by -swim.

Thanks! I imagine I'll send a PR for this.

> Good to learn about this tool - I've applied it, and then tweaked the result manually.

If you add a `.perltidyrc` file, then I can apply the same formatting to the snippets of code affected in PRs. https://metacpan.org/dist/Perl-Tidy/view/bin/perltidy#Using-a-.perltidyrc-command-file

For the parts you've tweaked manually, you can surround them with `perltidy` markers:

https://metacpan.org/dist/Perl-Tidy/view/bin/perltidy#Skipping-Selected-Sections-of-Code

That way subsequent runs of the formatter will ignore those lines and hopefully it will be easy to run `perltidy` on an entire file without undoing the manual changes.

oalders commented:
I just wanted to report that I had another swim this week. This time my watch did a better job of it, but I was able to trim almost 150m from the swim using the script. The `-smooth` option was able to trim it as part of this process.

It would be nice if I could filter out some other data, like if the reported heart rate is outside of a credible zone, it would be nice to remove it. That's one problem with having your watch track you under water. Another would be if I could give it start and end times and anything outside of that range could be ignored.

DDvO commented:
> > Good to learn about this tool - I've applied it, and then tweaked the result manually.
> 
> If you add a `.perltidyrc` file, then I can apply the same formatting to the snippets of code affected in PRs. https://metacpan.org/dist/Perl-Tidy/view/bin/perltidy#Using-a-.perltidyrc-command-file
> 
> For the parts you've tweaked manually, you can surround them with `perltidy` markers:
> 
> https://metacpan.org/dist/Perl-Tidy/view/bin/perltidy#Skipping-Selected-Sections-of-Code
> 
> That way subsequent runs of the formatter will ignore those lines and hopefully it will be easy to run `perltidy` on an entire file without undoing the manual changes.

Thanks for these further hints on `perltidy`.
Yet unfortunately I had not added a `.perltidyrc` file before using it,
and I fear that I'd have to mark many sections that I don't want to get touched.
Anyway, was good to run the tool once to clean up the formatting.

DDvO commented:
What I found much more important and finally did was to activate
```
use strict;
use warnings;
```
because this way undeclared and uninitialized variables are detected.
This way I found and fixed a number of small bugs.

DDvO commented:
> I just wanted to report that I had another swim this week. This time my watch did a better job of it, but I was able to trim almost 150m from the swim using the script. The `-smooth` option was able to trim it as part of this process.

Pleased to hear.

> It would be nice if I could filter out some other data, like if the reported heart rate is outside of a credible zone, it would be nice to remove it. That's one problem with having your watch track you under water. Another would be if I could give it start and end times and anything outside of that range could be ignored.

I've just added options supporting all this. For instance, one may use
```
-max hdop 21.5 -begin "2022-05-25 12:00" -end 10:05:22
```
to filter out points with HDOP values above 21.5 and before noon of May 25 and after today 10:05:22 AM.

DDvO commented:
Just added also a `-min` option and reports about the numbers of points filtered out for which reason.

oalders commented:
I just filtered another swim from yesterday down from 920 metres to 770 metres. So, this script is helpful as ever. 🎉 

Thank you for the latest improvements. I have a file in mind to test out the begin/end constraints.

It looks like I was really vague about filtering out data. A concrete example:

```xml
   <trkpt lat="43.3110340" lon="-79.7988760">
    <ele>74.8</ele>
    <time>2022-09-21T23:03:29Z</time>
    <extensions>
     <gpxtpx:TrackPointExtension>
      <gpxtpx:hr>0</gpxtpx:hr>
     </gpxtpx:TrackPointExtension>
    </extensions>
```

The `<gpxtpx:hr>0</gpxtpx:hr>` is essentially meaningless, but for some reason Strava doesn't deal well with unrealistic heart rate data. What I'd like to be able to do is toss out metadata that doesn't appear to be reasonable. I think there may be a good argument that that's beyond the scope of this script, but it would be so handy. To be clear I don't want to toss out `trkpt`, but if the script could toss out metadata for hr that wasn't within a certain range, that would make the data on the whole more accurate.

I could also write another script to do just the filtering, but I thought I would mention it.

oalders commented:
Looking at the source and the way the XML is generated, I think weeding out some of this data may be non-trivial. 🤔

oalders commented:
I just ran it on a GPX file that had originally reported a 3.8km race to be 6,887 meters. After running this and manually editing out the last few points I got it down to 3,779 meters. In reality my lines wouldn't have been that straight, but that's very close to the measured distance of the race and I'm quite happy with that.

https://www.strava.com/activities/7851096946

DDvO commented:
> Looking at the source and the way the XML is generated, I think weeding out some of this data may be non-trivial. thinking

The string processing capabilities of Perl are pretty strong, so don't worry.

DDvO commented:
> Thank you for the latest improvements. I have a file in mind to test out the begin/end constraints.

You are welcome.

> It looks like I was really vague about filtering out data. A concrete example:
> 
> ```
>    <trkpt lat="43.3110340" lon="-79.7988760">
>     <ele>74.8</ele>
>     <time>2022-09-21T23:03:29Z</time>
>     <extensions>
>      <gpxtpx:TrackPointExtension>
>       <gpxtpx:hr>0</gpxtpx:hr>
>      </gpxtpx:TrackPointExtension>
>     </extensions>
> ```
> 
> The `<gpxtpx:hr>0</gpxtpx:hr>` is essentially meaningless, but for some reason Strava doesn't deal well with unrealistic heart rate data. What I'd like to be able to do is toss out metadata that doesn't appear to be reasonable. I think there may be a good argument that that's beyond the scope of this script, but it would be so handy. To be clear I don't want to toss out `trkpt`, but if the script could toss out metadata for hr that wasn't within a certain range, that would make the data on the whole more accurate.

Got it - anyway, the `-max` and `-min` options were straightforward to implement and will be useful in other cases.
What you can already do is to prune all extension data using `-prune_exts`, but his would be too harsh.
No problem, I can quickly provide some options for selectively pruning out-of-range extensions.

DDvO commented:
Done.
Now for instance `-min_ext gpxtpx:hr 20 -max_ext gpxtpx:hr 200` should achieve what you need.

oalders commented:
> Now for instance -min_ext gpxtpx:hr 30 -max_ext gpxtpx:hr 300 should achieve what you need.

Yes, that's wonderful! It works beautifully. Thank you!

DDvO commented:
My pleasure.
I've just done some further minor improvements.

If you provide me feedback/corrections to the following suggested limit data for swimming:
```
    $MIN_TIMEDIFF          = 1;
    $MAX_SENSIBLE_SPEED    = 10;
    $MAX_SENSIBLE_ACCEL    = 1;
    $MAX_SENSIBLE_ELE_GAIN = 1;
    $MAX_SENSIBLE_ELE_LOSS = 1;
```
I can add the `-swim` option any time.
